pkg/kubevirt/core/core.go

```diff
@@ -0,0 +1,320 @@
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if p.client == nil {
+		kubevirtClient, err := kubevirtClient(secrets)
+		if err != nil {
+			return "", fmt.Errorf("failed to get kubevirt client: %v", err)
+		}
+		p.client = kubevirtClient
+	}
+
+	if err := p.client.Create(ctx, virtualMachine); err != nil {
+		return "", fmt.Errorf("failed to create vmi: %v", err)
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            userdataSecretName,
+			Namespace:       virtualMachine.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(virtualMachine, kubevirtv1.VirtualMachineGroupVersionKind)},
+		},
+		Data: map[string][]byte{"userdata": []byte(secrets.Data["userData"])},
+	}
+
+	if err := p.client.Create(ctx, secret); err != nil {
+		return "", fmt.Errorf("failed to create secret for userdata: %v", err)
+	}
+
+	// TODO(MQ): do we really need this? another approach is to return an empty provider id and leave it to the
+	// Get
+	return p.machineProviderID(secrets, machineName, providerSpec.Namespace)
+}
+
+func (p PluginSPIImpl) DeleteMachine(ctx context.Context, machineName, providerID string, providerSpec *api.KubeVirtProviderSpec, secrets *corev1.Secret) (foundProviderID string, err error) {
+	vm, err := p.getVM(secrets, machineName, providerSpec.Namespace)
+	if err != nil {
+		if clouderrors.IsMachineNotFoundError(err) {
+			klog.V(2).Infof("skip vm evicting, vm instance %s is not found", machineName)
+			return "", nil
+		}
+		return "", err
+	}
+
+	if vm != nil {
+		if p.client == nil {
+			p.client, err = kubevirtClient(secrets)
+			if err != nil {
+				return "", fmt.Errorf("failed to create kubevirt kubevirtClient: %v", err)
+			}
+		}
+
+		if err := p.client.Delete(context.Background(), vm); err != nil {
+			return "", fmt.Errorf("failed to delete vm %v: %v", machineName, err)
+		}
+		return encodeProviderID(string(vm.UID)), nil
+	}
+
+	return "", nil
+}
+
+func (p PluginSPIImpl) GetMachineStatus(ctx context.Context, machineName, providerID string, providerSpec *api.KubeVirtProviderSpec, secrets *corev1.Secret) (foundProviderID string, err error) {
+	return p.machineProviderID(secrets, machineName, providerSpec.Namespace)
+}
+
+func (p PluginSPIImpl) ListMachines(ctx context.Context, providerSpec *api.KubeVirtProviderSpec, secrets *corev1.Secret) (providerIDList map[string]string, err error) {
+	return p.listVMs(ctx, secrets)
+}
+
+func (p PluginSPIImpl) ShutDownMachine(ctx context.Context, machineName, providerID string, providerSpec *api.KubeVirtProviderSpec, secrets *corev1.Secret) (foundProviderID string, err error) {
+	virtualMachine, err := p.getVM(secrets, machineName, providerSpec.Namespace)
+	if err != nil {
+		return "", err
+	}
+	if virtualMachine != nil {
+		virtualMachine.Spec.Running = utilpointer.BoolPtr(false)
+		if err := p.client.Update(ctx, virtualMachine); err != nil {
+			return "", fmt.Errorf("failed to update machine running state: %v", err)
+		}
```

Comment:
But then you are actually handling the re-update inside the provider itself, where the requests to retry again from the reconciler itself not the cloud provider. Similar like what do you do in the machine creation. So when something goes wrong and re-enqueue the request again. Or do you have some certain criteria for re-enqueuing requests?

Reply:
As discussed in the meeting, both approaches are valid. I tend to slightly prefer handing `Conflict` errors in the way recommended above, since they are very common. This is also what's typically done in the Gardener codebase.